tests/test_bern.py

```diff
@@ -0,0 +1,167 @@
+
+    MC = MeshConstant(msh)
+    t = MC.Constant(0.0)
+    dt = MC.Constant(2.0 / N)
+
+    uexact = exp(-t) * cos(pi * x)**2
+    rhs = expand_derivatives(diff(uexact, t)) - div(grad(uexact))
+
+    u = project(uexact, V)
+
+    v = TestFunction(V)
+
+    F = (inner(Dt(u), v) * dx + inner(grad(u), grad(v)) * dx
+         - inner(rhs, v) * dx)
+
+    bc = DirichletBC(V, uexact, "on_boundary")
+
+    stepper = TimeStepper(F, butcher_tableau, t, dt, u,
+                          bcs=bc, solver_parameters=solver_parameters,
+                          **kwargs)
+
+    while (float(t) < 1.0):
+        if (float(t) + float(dt) > 1.0):
+            dt.assign(1.0 - float(t))
+        stepper.advance()
+        t.assign(float(t) + float(dt))
+
+    return errornorm(uexact, u) / norm(uexact)
+
+
+def mixed_heat(n, deg, butcher_tableau, solver_parameters,
+               **kwargs):
+    N = 2**n
+    msh = UnitSquareMesh(N, N)
+
+    V = FunctionSpace(msh, "RT", deg)
+    el_type = "Bernstein" if deg > 1 else "DG"
+    W = FunctionSpace(msh, el_type, deg-1)
+
+    Z = V * W
+
+    x, y = SpatialCoordinate(msh)
+
+    MC = MeshConstant(msh)
+    t = MC.Constant(0.0)
+    dt = MC.Constant(1.0 / N)
+
+    pexact = (cos(pi * x) * cos(pi * y))**2 * exp(-t)
+    uexact = -grad(pexact)
+
+    up = Function(Z)
+    u, p = split(up)
+
+    v, w = TestFunctions(Z)
+
+    n = FacetNormal(msh)
+
+    rhs = expand_derivatives(diff(pexact, t) + div(uexact))
+
+    F = (inner(Dt(p), w) * dx
+         + inner(div(u), w) * dx
+         - inner(rhs, w) * dx
+         + inner(u, v) * dx
+         - inner(p, div(v)) * dx
+         + inner(pexact, dot(v, n)) * ds)
+
+    up.subfunctions[1].project(pexact)
+
+    stepper = TimeStepper(F, butcher_tableau, t, dt, up,
+                          solver_parameters=solver_parameters,
+                          **kwargs)
+
+    while (float(t) < 1.0):
+        if (float(t) + float(dt) > 1.0):
+            dt.assign(1.0 - float(t))
+        stepper.advance()
+        t.assign(float(t) + float(dt))
+
+    u, p = up.subfunctions
+    erru = errornorm(uexact, u) / norm(uexact)
+    errp = errornorm(pexact, p) / norm(pexact)
+    return erru + errp
+
+
+@pytest.mark.parametrize('butcher_tableau', [RadauIIA(i) for i in (1, 2)])
+def test_mixed_heat_bern(butcher_tableau):
+    deg = 1
+    kwargs = {"stage_type": "value",
+              "basis_type": "Bernstein",
+              "solver_parameters": lu_params}
+    diff = np.array([mixed_heat(i, deg, butcher_tableau, **kwargs) for i in range(2, 4)])
+    print(diff)
+    conv = np.log2(diff[:-1] / diff[1:])
+    print(conv)
+    assert (conv > (deg-0.1)).all()
+
+
+@pytest.mark.parametrize('num_stages', (2, 3))
+@pytest.mark.parametrize('bounds_type', ('stage', 'last_stage'))
+@pytest.mark.parametrize('basis_type', ('Bernstein', None))
+def test_mixed_heat_bern_bounds(num_stages, bounds_type, basis_type):
+    deg = 1
+    bounds = (bounds_type, (None, 0), (None, None))
```

Comment:
This remains quite cryptic to me.  I think the right thing to do is have a fully commented demo that shows how to set bounds and use them...

Reply:
Good point, I'll double check order against Firedrake/PETSc.  We can probably unpack the `None` option.  We will have a documented demo in before we try to merge.

Reply:
Yes, the order is fixed and we are using lower, upper now.

Reply:
And there are fewer options now  -- we just have `upper` and `lower` (like for the NLVS).  If you want to constrain one piece of a mixed space, it's on you to set the bounds on the other sub functions to be INFINITY or NINFINITY.

Reply:
This is pretty fully reworked.  See updated tests and also john's committed demo.